### PR TITLE
Expand previously-opened-element check to include all currently-open elements

### LINKED ISSUE #3: Potential bug: implicitly-closed tag lookup only considers a single previous element
Hi @willforde - I don't know for certain whether this is a bug - it also looks like it could've been a performance-related decision, but either way, I figured it'd be worth asking about.

There's some code at https://github.com/willforde/python-htmlement/blob/20b365a06605a741944791937bafe21b6ab2fbce/htmlement.py#L366-L368 that detects "unexpected" closing tags (in other words, we've found a closed tag, but it's not for the element currently being processed).

At the moment, that check only inspects the previous element (second-from-last in the element stack/list) for a match.

There aren't many negative side-effects of this as far as I can tell, but it does allow for more text than expected to be gathered into the `_data` buffer some cases, and I wondered whether the choice of array index check is intentional?

I'll offer a pull request with an example text-handling problem that can occur, and a possible alternative implementation for the previously-opened-tag condition.

### PR DESCRIPTION
**Pull request by jayaddison**
Could resolve #3 - also has the potential to reduce performance, though.

### CODE REVIEW

File: tests/test_module.py [hunk 1]
```diff
@@ -170,6 +170,22 @@ def test_entity_hex_euro_fail():
...
+
+
+def test_text_iterator():
+    html = "<html><body>sample text content</body></html>"
+    root = quick_parsehtml(html)
+    body = root.find(".//body")
+    assert "".join(body.itertext()) == "sample text content"
+
+def test_text_iterator_unclosed_tag():
+    html = "<html><body><span>hello <div>to <div>the <div>world!</span></body><footer>unrelated</footer></html>"
```
Comment (jayaddison) [inline tests/test_module.py]: Note that this test coverage is intentionally designed not to make any assumptions about the tree/node structure built in-memory by the parser.. because I'm not sure about the implications yet.

(for example: `div<div<div>>>` (nested nodes) is one possible in-memory representation, and `div<>div<>div<>` (sibling nodes) is another) -- either of those produce the same `itertext` results in _this_ test case, but it's possible to imagine representations where [descendent text content](https://dom.spec.whatwg.org/#concept-descendant-text-content) would be different depending on the parse tree)

Comment (jayaddison) [reply tests/test_module.py]: ...and it turns out, the tree structure produced by `lxml` when 'implied end tags' are generated depends on the elements encountered! (by design)

A series of unclosed `div` elements will turn into nested nodes, whereas a series of unclosed `li` items will result in sibling nodes.

It's all spec'd in the whatwg HTML5 spec, but it's kinda ~~complicated~~ involved (clarify: not complicated.. a lot of rules, though).  Both the current node and the previous node are relevant when deciding how to emit implied closing tags, and there are custom rules for various different elements.

Uncertain what to do here - I like that this library is simple and fast.  The reason I'm contributing is to see whether ScrapingHub's `extruct` library could use `htmlement` instead of `lxml`, allowing me to build pure-Python containers for `openculinary`.

Recent context here: https://github.com/scrapinghub/extruct/pull/163#issuecomment-1287166299 (and longer story earlier in that same issue)

In short: some of the test cases in `extruct` relate to tree-order traversal of parsed HTML nodes for text content retrieval, which is what led me to all this.

**Comment by willforde** [comment]
Hello. This is a good change, thanks. There shouldn't be much of a performance hit with this change, as it shouldn't be hit too often.

 I didn't plan for more than one malformed end tag, that was a bit short sighted on my part. Thanks again.

**Comment by jayaddison** [comment]
You're welcome, glad to contribute - this seemed to make one small step towards migration away from `lxml` in [`extruct`](https://github.com/scrapinghub/extruct.git/) (a library that `openculinary` uses, and that I'd like to see freed of non-Python dependencies).

As commented above, I'm not sure whether that migration is practical or straightforward after all -- not a fault of `htmlement` - seems tricky to get the text fragment ordering logic for HTML5 correct.

File: tests/test_module.py (hunk 1 above)
Comment (willforde) [reply tests/test_module.py]: Yeah that is a bit of an issue alright. While Htmlement's approach is fast, it's also very naive, deliberately so. 
I can add more logic to catch some of the more common closing tag issues, but it would never be fully compliant.
